app/user.go

```diff
@@ -1215,6 +1215,20 @@ func (a *App) PatchUser(userID string, patch *model.UserPatch, asAdmin bool) (*m
 		return nil, err
 	}
 
+	if patch.Username != nil && *patch.Username != user.Username {
+		u, err := a.GetUserByUsername(*patch.Username)
+		if err == nil || u != nil {
+			return nil, model.NewAppError("patchUser", "app.user.save.username_exists.app_error", nil, "", http.StatusBadRequest)
+		}
+	}
+
+	if patch.Email != nil && *patch.Email != user.Email {
+		u, err := a.GetUserByEmail(*patch.Email)
+		if err == nil || u != nil {
+			return nil, model.NewAppError("patchUser", "app.user.save.email_exists.app_error", nil, "", http.StatusBadRequest)
+		}
+	}
+
```

Comment:
I think applying the check manually is not needed here. The relevant constraints are already handled in the DB layer. And the email check is already done in `a.UpdateUser`:

```go
if _, appErr := a.GetUserByEmail(user.Email); appErr == nil {
			return nil, model.NewAppError("UpdateUser", "store.sql_user.update.email_taken.app_error", nil, "user_id="+user.Id, http.StatusBadRequest)
		}
```

What's incorrect is our tagging of the error to the translation string. We already check for clashes here:
```go
if IsUniqueConstraintError(err, []string{"Email", "users_email_key", "idx_users_email_unique"}) {
			return nil, store.NewErrInvalidInput("User", "id", user.Id)
		}
		if IsUniqueConstraintError(err, []string{"Username", "users_username_key", "idx_users_username_unique"}) {
			return nil, store.NewErrInvalidInput("User", "id", user.Id)
		}
```

But at the call site, we dump the invalid input as a "cannot find" error:

```
case errors.As(err, &invErr):
			return nil, model.NewAppError("UpdateUser", "app.user.update.find.app_error", nil, invErr.Error(), http.StatusBadRequest
```

I think it might be better to return `store.NewErrConflict` for those two cases, and handle the case separately and return the correct tx id.

Reply:
Wow, thanks! Going to update accordingly 💪